File: src/main/java/at/flauschigesalex/minecraftLibrary/bukkit/ComponentManager.java
package at.flauschigesalex.minecraftLibrary.bukkit;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("unused")
public final class ComponentManager {

    public static ArrayList<Component> spliterator(@NotNull Component component) {
        return spliterator(component, new ArrayList<>());
    }

    private static ArrayList<Component> spliterator(@NotNull Component component, @NotNull ArrayList<Component> list) {
        Component base = list.isEmpty() ? Component.empty() : list.getLast();
        component = component.replaceText(builder -> builder.match("<br>").replacement("<newLine>"));

        final Component withoutChild = component.children(List.of());
        if (withoutChild instanceof TextComponent textComponent && textComponent.content().contains("\n")) {
            for (final String string : textComponent.content().split("\n")) {
                if (!string.isBlank() && !string.isEmpty())
                    base = base.append(Component.text(string).style(textComponent.style()));

                list.add(base);
                base = Component.empty();
            }
        } else if (withoutChild.equals(Component.newline())) {
            list.add(base);
            base = Component.empty();
        } else base = base.append(withoutChild);

        for (final Component child : component.children())
            spliterator(child, list);

        if (!base.equals(Component.empty()))
            list.add(base);

        return list;
    }

    public static Component compromiser(final List<Component> list) {
        if (list == null || list.isEmpty())
            return Component.empty();

        Component base = list.getFirst();
        for (int i = 1; i < list.size(); i++) {
            base = base.append(list.get(i));
        }

        return base;
    }
}

File: src/main/java/at/flauschigesalex/minecraftLibrary/bukkit/utils/ComponentManager.kt
package at.flauschigesalex.minecraftLibrary.bukkit.utils

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.TextComponent

@Suppress("unused")
object ComponentManager {
    @JvmStatic fun spliterator(component: Component): ArrayList<Component> {
        return spliterator(component, ArrayList())
    }

    private fun <E: Any> ArrayList<E>.setLast(value: E): Pair<ArrayList<E>, E?> {
        if (this.isEmpty()) {
            this.add(value)
            return Pair(this, null)
        }

        val old = this.lastOrNull()
        this[this.size -1] = value

        return Pair(this, old)
    }

    private fun spliterator(component: Component, list: ArrayList<Component>): ArrayList<Component> {
        val base = list.lastOrNull() ?: Component.empty()
        val noChildren = component.children(listOf())

        if (noChildren is TextComponent && noChildren.content() == "\n") {
            list.setLast(base)
            list.add(Component.empty())
        } else {
            list.setLast(base.append(noChildren))
        }

        component.children().forEach {
            spliterator(it, list)
        }

        return list
    }

    @JvmStatic fun compromiser(list: List<Component?>?): Component {
        if (list.isNullOrEmpty())
            return Component.empty()

        var base: Component = list.first()!!
        for (i in 1 until list.size) {
            base = base.append(list[i]!!)
        }

        return base
    }
}
